Migrate to futures 0.3

derekdreery opened the pull request:
This patch moves all crates & tests over to v0.3 (including doc tests). Closes #97 

I need to check if there is any other documentation that needs to be updated.

Also there are a few style decisions to be made as regards to async style: for example I've done

```rust
spawn_local(async {
    let mut interval = IntervalStream::new(1_000);
    loop {
        interval.next().await; // I'm discarding the result as I know it will go on for ever (might not be true if someone cancels the interval from js or using its i32 id)
        // Do something every minute
    }
})

derekdreery commented:
closes #97

derekdreery commented:
CI is failing because it is trying to use an older stable rust. Is there anything I need to do the .asure file?

Pauan commented:
(Sorry for the delay on this, it's a meaty PR)

Pauan commented on crates/events/README.md:
```diff
@@ -20,6 +20,6 @@
 </div>
 
-Using event listeners with [`web-sys`](https://crates.io/crates/web-sys) is hard! This crate provides an [`EventListener`](https://docs.rs/gloo-events/0.1.0/gloo_events/struct.EventListener.html) type which makes it easy!
+Using event listeners with [`web-sys`](https://crates.io/crates/web-sys) is hard! This crate provides an [`EventListener`](struct.EventListener.html) type which makes it easy!
```
These changes seem wrong, since it will cause the URL to point to `https://crates.io/crates/gloo-events/struct.EventListener.html`, not `https://docs.rs/gloo-events/0.1.1/gloo_events/struct.EventListener.html` like it should.

However, I would be in favor of changing the URL to use `^0.1.0` instead, so it will point to the latest version.

derekdreery replied:
Hmm, I don't remember making this change. Are the READMEs auto-generated? Anyway it's no trouble to revert it. if necessary.

derekdreery replied:
I've changed it to `^0.1.0` like you suggested.

Pauan commented on crates/events/src/lib.rs:
```diff
@@ -214,14 +215,17 @@ thread_local! {
-///     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
-///         self.receiver.poll().map_err(|_| unreachable!())
+///     fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+///         self.project_receiver().poll_next(cx)
```
This doesn't need `project_receiver` or unsafe code, it can just do this:

```rust
Pin::new(&mut self.receiver).poll_next(cx)
```

derekdreery replied:
I feel like I need to get more of an intuition for `Pin`s. Does this work because `UnboundedReceiver<()>` is `Unpin`?

Pauan replied:
Yes, `Unpin` is similar to other auto-traits like `Sync` or `Send`: if all of the fields of your struct are `Unpin`, then the struct as a whole is `Unpin`. Almost all types in Rust are `Unpin`.

And if a struct is `Unpin` then you can safely access its fields (because of [this `DerefMut` impl](https://doc.rust-lang.org/nightly/std/pin/struct.Pin.html#impl-DerefMut)), and you can also safely use `Pin::new(&mut foo)` (because [`new` has an `Unpin` bound](https://doc.rust-lang.org/nightly/std/pin/struct.Pin.html#method.new)).

So the only time you need to use unsafe code is when one (or more) of your fields is *not* `Unpin`. This usually only happens in Future/Stream *combinators*, but not with leaf types like `OnClick`.

Or to put it another way: the burden of unsafe code is only for the creators of Future/Stream combinators, and not anybody else.

Pauan commented on crates/events/tests/web.rs:
```diff
@@ -38,7 +37,7 @@ where
 #[derive(Clone)]
 struct Sender<A> {
-    sender: mpsc::UnboundedSender<Result<A, JsValue>>,
+    sender: futures::channel::mpsc::UnboundedSender<Result<A, JsValue>>,
```
I'd rather this continue to use the import at the top for brevity.

Pauan commented on crates/events/tests/web.rs:
```diff
@@ -50,29 +49,29 @@ impl<A> Sender<A> {
-    mpsc(|sender| {
+#[wasm_bindgen_test]
+async fn new_with_options() {
+    let results = mpsc::<(), _>(|sender| {
```
Is this type annotation needed? If so, I would prefer if it were `let results: Vec<()> = mpsc(|sender| {` instead.

Pauan commented on crates/events/tests/web.rs:
```diff
@@ -93,13 +92,13 @@ fn new_with_options() -> impl Future<Item = (), Error = JsValue> {
-    })
-    .and_then(|results| is(results, vec![(), ()]))
+    }).await;
+    is(results, vec![(), ()]).unwrap_throw();
```
This can be replaced with `assert_eq`

Pauan commented on crates/events/tests/web.rs:
```diff
@@ -50,29 +49,29 @@ impl<A> Sender<A> {
 
     receiver
-        .then(|x| match x {
+        .map(|x| match x {
```
This doesn't seem right, it's not handling the `Result` properly. It should be something like this instead:

```rust
receiver.try_collect().await
```

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -1,8 +1,11 @@
+use std::task::{Poll, Context};
+use std::pin::Pin;
+use std::future::Future;
 use super::sys::*;
```
Could you move the super import to the top? I like to keep imports roughly grouped by category.

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -1,8 +1,11 @@
 use super::sys::*;
-use futures::prelude::*;
-use futures::sync::mpsc;
+use futures::stream::Stream;
```
We should be using individual crates like `futures_core` rather than the blanket `futures`.

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -24,27 +27,20 @@ use wasm_bindgen_futures::JsFuture;
-///         .and_then(|(who, other)| {
-///             // `timeout_a` should have won this race.
-///             assert_eq!(who, "a");
+/// let timeout_a = TimeoutFuture::new(1_000);
```
Can these be moved inside of the `async` block?

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -24,27 +27,20 @@ use wasm_bindgen_futures::JsFuture;
+///             drop(b);
+///         }
+///         future::Either::Right((a, val)) => {
+///             panic!("timeout_a should have won this race")
```
Nit: `;` at the end of the line.

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -24,27 +27,20 @@ use wasm_bindgen_futures::JsFuture;
 ///
+/// wasm_bindgen_futures::spawn_local(async {
+///     match future::select(timeout_a, timeout_b).await {
+///         future::Either::Left((val, b)) => {
```
I'd prefer if these were all imported with `use`

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -75,11 +71,10 @@ impl TimeoutFuture {
-    ///         // Do stuff after one second...
-    ///     })
-    /// );
+    /// wasm_bindgen_futures::spawn_local(async {
```
Ditto, this should be imported with `use`

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -90,18 +85,25 @@ impl TimeoutFuture {
-            // We only ever `resolve` the promise, never reject it.
-            Err(_) => wasm_bindgen::throw_str("unreachable"),
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        match self.project_inner().poll(cx) {
```
Similar to above, this doesn't need `unsafe` code, it can just use `Pin::new(&mut self.inner).poll(cx)`

Also, this entire thing can actually be massively simplified by using `set_timeout` directly (not using `JsFuture`). But we can save that for another PR if you like.

derekdreery replied:
If we use `set_timeout`, should the timer start when the future is created (`TimeoutFuture::new`), or when it is actually `poll`ed?

derekdreery replied:
The current impl uses the `TimeoutFuture::new` tick to initiate the timeout. This results in a slightly more complex state machine.

Pauan replied:
That's a really good question. I think there's reasonable arguments for both behaviors, but having it activate when polling is more logically consistent with `async`.

In any case we can discuss that in a new PR.

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -136,13 +138,13 @@ impl IntervalStream {
-    /// );
+    /// wasm_bindgen_futures::spawn_local(async {
+    ///     let mut interval = IntervalStream::new(1_000);
+    ///     loop {
```
This should use `StreamExt::for_each`, like this:

```rust
spawn_local(async {
    IntervalStream::new(1_000).for_each(|_| {
        // Do stuff every one second...
    }).await;
});
```

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -169,16 +177,15 @@ impl Drop for IntervalStream {
         }
 
-        self.inner.poll()
+        self.project_inner().poll_next(cx)
```
Similar to above, this can just be `Pin::new(&mut self.inner).poll_next(cx)`

Pauan commented on crates/timers/src/lib.rs:
```diff
@@ -38,14 +38,12 @@ use futures::prelude::*;
 // the timeout would be cancelled with `clearTimeout`.
-spawn_local(timeout);
+spawn_local(async {
+    TimeoutFuture::new(1_000_).await;
```
Nit: `_` is not needed at the end.

Pauan commented on crates/timers/tests/web.rs:
```diff
@@ -44,72 +45,62 @@ fn timeout_cancel() -> impl Future<Item = (), Error = wasm_bindgen::JsValue> {
         move |_| {
             assert_eq!(cell.get(), false);
-            1
+            1 as u32
```
Why does this need a cast? Assuming the type inference is failing, it can use `1u32` instead.

derekdreery commented:
I've addressed all your feedback. Mostly it was just fixing things, but in the case of the future impl for timeouts I'd appreciate it if you'd have a look.

derekdreery commented:
I added an example just to convince myself that intervals are working. I'm happy to remove it if you prefer.

derekdreery commented:
Friendly ping :)

heilhead commented:
Any news on this? Excited to try out!

derekdreery commented on crates/timers/src/future.rs:
```diff
@@ -71,37 +54,32 @@ impl TimeoutFuture {
+        match Future::poll(Pin::new(&mut self.rx), cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Ok(t)) => Poll::Ready(t),
+            Poll::Ready(Err(e)) => panic!("{}", e),
```
Note to self - I need to put the comment back in here to explain why panicking is ok (that branch is unreachable)

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -71,37 +54,32 @@ impl TimeoutFuture {
+        let (tx, rx) = oneshot::channel();
+        let inner = Timeout::new(millis, move || {
+            // if the receiver was dropped we do nothing.
+            let _ = tx.send(());
```
I think this should `unwrap_throw`, just as an extra precaution.

derekdreery replied:
Yep you're right if the receiver is dropped then the whole thing is dropped so this closure should never run.

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -71,37 +54,32 @@ impl TimeoutFuture {
-            // We only ever `resolve` the promise, never reject it.
-            Err(_) => wasm_bindgen::throw_str("unreachable"),
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        match Future::poll(Pin::new(&mut self.rx), cx) {
```
I'd rather this be written as `Pin::new(&mut self.rx).poll(cx).map(|t| t.unwrap_throw())`

Pauan commented on crates/timers/src/future.rs:
```diff
@@ -132,53 +108,31 @@ impl IntervalStream {
-}
+        let inner = Interval::new(millis, move || {
+            // if the receiver was dropped we do nothing.
+            let _ = sender.unbounded_send(());
```
Similarly, I like `unwrap_throw` for extra safety.

Pauan commented on crates/timers/tests/web.rs:
```diff
@@ -68,48 +66,38 @@ fn timeout_future_cancel() -> impl Future<Item = (), Error = wasm_bindgen::JsVal
         move |_| {
             cell.set(true);
-            2
+            2 as u32
```
Should be `2u32`

derekdreery replied:
I think the type system can infer this now. However, I can be explicit if you prefer.

Pauan commented on examples/clock/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "clock"
+version = "0.1.0"
+authors = ["Richard Dodd <richard.o.dodd@gmail.com>"]
```
This needs to be `authors = ["Rust and WebAssembly Working Group"]`

Pauan commented on examples/clock/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+    "Node",
+]
+
+[workspace]
```
Is there a reason for this `[workspace]`?

derekdreery replied:
It's here so that the root `Cargo.toml` doesn't complain about it not being in the root workspace. (try removing it and building to see the problem).

Pauan replied:
Ah, okay.

Pauan commented on examples/clock/src/lib.rs:
```diff
@@ -0,0 +1,30 @@
+#[wasm_bindgen(start)]
+pub fn main() {
+    console_error_panic_hook::set_once();
+    let document = web_sys::window().unwrap_throw().document().unwrap_throw();
```
Nit: could you add some empty lines in here to spread things out so they look nicer?

derekdreery replied:
I've tried to group things together in a sensible way, could you have a look and tell me what you think.

Pauan replied:
I'd personally prefer a newline between every statement, but I guess it doesn't matter too much.

derekdreery replied:
@Pauan I've added newlines - is this what you mean?

Pauan replied:
Looks great, thanks!

Pauan commented on examples/clock/src/lib.rs:
```diff
@@ -0,0 +1,30 @@
+}
+
+fn render_date(el: &web_sys::Element) {
+    let date = chrono::Local::now();
```
Nit: same here, a few newlines for readability.

derekdreery commented on crates/events/src/lib.rs:
```diff
@@ -218,10 +223,9 @@ thread_local! {
 ///
-///     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
-///         self.receiver.poll().map_err(|_| unreachable!())
+///     fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
```
I wonder why my testing didn't catch this.

Pauan replied:
Yes, that's odd. Maybe it's not running it? Try putting in something that's an obvious error (like an undefined variable) and see if it catches it.

derekdreery commented:
Ready for review again.

heilhead commented:
I'm getting a bunch of errors trying to build now (`cargo build --target wasm32-unknown-unknown --features futures`):
```
error[E0599]: no method named `unwrap_throw` found for type `std::result::Result<(), ()>` in the current scope
  --> crates/timers/src/future.rs:69:25
   |
69 |             tx.send(()).unwrap_throw();
   |                         ^^^^^^^^^^^^ help: there is a method with a similar name: `unwrap_err`
   |
   = help: items from traits can only be used if the trait is in scope
   = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
           `use wasm_bindgen::UnwrapThrowExt;`

error[E0599]: no method named `unwrap_throw` found for type `std::result::Result<(), futures_channel::oneshot::Canceled>` in the current scope
  --> crates/timers/src/future.rs:79:60
   |
79 |         Future::poll(Pin::new(&mut self.rx), cx).map(|t| t.unwrap_throw())
   |                                                            ^^^^^^^^^^^^ help: there is a method with a similar name: `unwrap_err`
   |
   = help: items from traits can only be used if the trait is in scope
   = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
           `use wasm_bindgen::UnwrapThrowExt;`

error[E0599]: no method named `unwrap_throw` found for type `std::result::Result<(), futures_channel::mpsc::TrySendError<()>>` in the current scope
   --> crates/timers/src/future.rs:121:39
    |
121 |             sender.unbounded_send(()).unwrap_throw();
    |                                       ^^^^^^^^^^^^ help: there is a method with a similar name: `unwrap_err`
    |
    = help: items from traits can only be used if the trait is in scope
    = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
            `use wasm_bindgen::UnwrapThrowExt;`

error: aborting due to 3 previous errors
```

Pauan commented:
@heilhead Yeah, it needs to add in a `use wasm_bindgen::prelude::*` in order to fix that.

derekdreery commented:
Thanks for spotting the build problem. @Pauan is there a way I can fix the azure script so it will build this patch? The problem is that it is using rustc 1.38 which is the version before async/await. I don't know why it pulls this version in as the ci scripts say to use latest stable.

Pauan commented:
@derekdreery Okay, I think I fixed the CI stuff, could you rebase and push?

Pauan commented:
Looks good, thanks a lot for this!

zimond commented:
@Pauan can you please pblish a new version of `gloo-timers` with this update?

Pauan commented:
@zimond Yeah, I need to fix https://github.com/rustwasm/gloo/issues/96 first before I publish.

Pauan commented:
@zimond Version `0.2.0` is out now!

zimond commented:
haha, thanks
